#469 Mapping Object Graphs with Cycles

pascalgn opened the pull request:
Fixes issue #469

agudian commented:
@pascalgn: awesome! Great that you're approaching this issue.

I'm pretty sure there are some other use cases for handling return values of before/after methods as well. But it would defintely need to work without any casts or instanceof in the generated code. In particular, I think in this very case you're demonstrating in the test, this wouldn't even be necessary as you should be able to specify your callback methods with generics, e.g. `public static <T> T getInstance(Object source, @TargetType Class<T> targetType)`. MapStruct should already be able to check that it would be a valid method to be assignable to the target bean type.

However, when we start allowing `@BeforeMapping` methods to return something, then why wouldn't we also allow an `@AfterMapping` method to return a value? For example, it could be a proxied or decorated version of the target object, or a JPA managed instance of it that should be returned instead of the originally created instance. Plus, we currently allow multiple before/after mapping methods to be specified and we generate calls to all of the candidates that match. So your PR would have to be able to deal with this as well.

And once that is in place, this feature could be used to handle cycles in the bean graph as you are demonstrating it. But I'm not sure if that is really what we should advertise as our solution to this problem. I'm actually more fond of explicitly passing a `MappingContext` (or whatever) into the mapping method explicitly that holds the state, we would pass that one down the call hierarchy and do the required calls on that context.

The way I see it is that we either focus this PR on allowing and properly handling non-void before/after methods (with all the strings attached to it) and add the cycle-detection as an example use case until we have a proper solution in place, or we instead try to come up with an API design for a `MappingContext` (maybe keeping it open for other possible use cases, such as conditional mapping of properties) and go down that road.

I think at some point we should have both features and both variants would get you to your solution. So I would leave it up to you and we will be happy with whatever road you choose and we'll support you with the PR either way.

What do you think, does that make sense for you?

pascalgn commented:
Thanks for the feedback!

1. Yes, you are right. My motivation was to always call the before method, but I had another look at the implementation and saw that before/after-mapping methods are only called if their parameters match. Therefore it makes perfect sense to only call the methods if the return values match, too. I have corrected this issue in the pull request, there are no more casts necessary.
2. At the time of writing the patch, I couldn't think of a good use case for @AfterMapping methods returning a mapped instance. However, your example makes perfect sense, so I have changed my patch accordingly.
3. I agree, having some sort of `MappingContext` would be a cleaner solution. However, using the before/after-mapping method approach was much easier, which is why I chose to implement it this way.

I have also updated the documentation. It no longer says that before/after-mapping methods must return void and I have added your example of an @AfterMapping methods that attaches entity to the persistence context.

Again, thanks for the feedback and if you have any other remarks/comments, just let me know!

----

Regarding the `MappingContext` solution: When implementing this patch, I also thought about introducing an interface to be used as a context. However, I did not continue this way, because that would require to have MapStruct as a runtime dependency. That would not have been a big problem for me, but I would imagine that some people use MapStruct because the generated code doesn't need any runtime dependencies; the solution with an interface might be a deal breaker for them.

However, what one might do is to kind of implement an interface using annotations. That is, instead of using `implements MappingContext`, users would just annotate their class and the respective methods:

    @MappingContext
    public class MyContext {
        @BeforeMapping
        public void beforeMapping() {
            // init context
        }

        @AfterMapping
        public void afterMapping() {
            // destroy context
        }
    }

That way, the MapStruct dependency would not be needed during runtime.

The main problem with the context solution (whether using an interface or annotations) is that the context needs to be passed in to all the mapping methods: Parameter mapping might get more complicated and there would need to be checks to make sure the context classes match. Cases must be considered where a user specifies a mapping method both with and without a mapping contect, etc. So it is certainly possible, but it is a bit more work.

AntwaneB commented:
Hello,

Any updates? Do you know when it will be merged if it is going to be?

Thanks :)

agudian commented:
We're getting there, it's just a matter of time for us to properly digest the changes (we concentrated our efforts on the way to 1.1.0.Final the last couple of weeks).
But the PR is in a good shape so I think we can get over with it soonish, I hope. Hang in there!
If you like, you could mere this on top of a for of the current master, build your own little release and play around with it a bit to see if it fits your needs, @AntwaneB. Perhaps you have some early feedback for us or @pascalgn even before we merge it 😃

AntwaneB commented:
Great, thanks for the update :)

If I have the time to work on this part of my application until the release I'll make sure to test it and let you know of the results! :)

Thanks.

agudian commented on processor/src/test/java/org/mapstruct/ap/test/bugs/_469/Attribute.java:
```diff
@@ -0,0 +1,64 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.test.bugs._469;
```
Could you please place the tests in a package somewhere below `org.mapstruct.ap.test.callbacks`, e.g. `...callbacks.returning`. For features and improvements, we try to name the packages and the tests according to that feature.

agudian commented on processor/src/test/java/org/mapstruct/ap/test/bugs/_469/Attribute.java:
```diff
@@ -0,0 +1,64 @@
+ */
+package org.mapstruct.ap.test.bugs._469;
+
+public class Attribute {
```
Please add an `@author` tag with your name to all of your newly created classes and take the credit you deserve 👍

agudian commented on processor/src/test/java/org/mapstruct/ap/test/bugs/_469/NodeMapperContext.java:
```diff
@@ -0,0 +1,99 @@
+    private static final List<ContextListener> LISTENERS = new CopyOnWriteArrayList<ContextListener>();
+
+    private static final ThreadLocal<Integer> LEVEL = new ThreadLocal<Integer>();
+    private static final ThreadLocal<Map<Object, Object>> MAPPING = new ThreadLocal<Map<Object, Object>>();
```
Please add some documentation on what parts of this class are for test-inspection (the listeners and the level, I guess?) and what is demonstrating the cycle-detection.

pascalgn commented:
Thank you for the feedback! I have just pushed your suggested edits. Just let me know if you have any other suggestions!

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/LifecycleCallbackFactory.java:
```diff
@@ -125,8 +141,9 @@ private LifecycleCallbackFactory() {
-
+                && callback.matches( extractSourceTypes( parameterAssignments ), method.getResultType() )
+                && ( callback.getReturnType().isVoid() || callback.getReturnType().isTypeVar()
+                    || callback.getReturnType().isAssignableTo( method.getResultType() ) ) ) {
```
I would extract the or checks to some method instead of doing everything in the `if`.

agudian replied:
You'll also have to check if `method` has a non-void `returnType` (`!method.getReturnType().isVoid()`). Plus, I think in the assignable-to check, should also take `method.getReturnType()` into account instead of the `resultType` (the `resultType` may be more specific than `returnType` for certain mapping methods with `@MappingTarget`, but in most cases both will be the same).

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/LifecycleCallbackMethodReference.java:
```diff
@@ -37,19 +37,52 @@
+        }
+
+        if ( hasReturnType() ) {
+            String possibleName = Introspector.decapitalize( "target" );
```
You can directly pass the `target` to the `Strings#getSaveVariableName`, as it is already lower case

agudian commented:
Dear Jenkins, this pr is ok to test :)

filiphr commented on processor/src/main/resources/org/mapstruct/ap/internal/model/LifecycleCallbackMethodReference.ftl:
```diff
@@ -18,4 +18,7 @@
 
 -->
-<#if declaringType??>${instanceVariableName}.</#if>${name}(<#list parameterAssignments as param> <#if param.targetType><@includeModel object=ext.targetType raw=true/>.class<#elseif param.mappingTarget>${ext.targetBeanName}<#else>${param.name}</#if><#if param_has_next>,<#else> </#if></#list>);
+<#if hasReturnType()>${mappingTargetType} ${targetVariableName} = </#if><#if declaringType??>${instanceVariableName}.</#if>${name}(<#list parameterAssignments as param> <#if param.targetType><@includeModel object=ext.targetType raw=true/>.class<#elseif param.mappingTarget>${ext.targetBeanName}<#else>${param.name}</#if><#if param_has_next>,<#else> </#if></#list>);
```
Use `<@includeModel object=mappingTargetType/>` instead of `${mappingTargetType}`. 
The second one will create something like `java.util.List<org.mapstruct.ap.test.callbacks.returning.AttributeDto> target` and the first one creates `List<AttributeDto>`. Using the model one uses the `Type.ftl`, whereas the other one uses the `Type#toString()`

As a side note for us. We can maybe use `<@compress single_line=true>` here to have a more readable template 😄

agudian replied:
This seems to give some good results and is slightly more readable:

```
<@compress single_line=true>
    <#if hasReturnType()>
        <@includeModel object=mappingTargetType /> ${targetVariableName} =
    </#if>
    <#if declaringType??>${instanceVariableName}.</#if>${name}(<#list parameterAssignments as param> <#if param.targetType><@includeModel object=ext.targetType raw=true/>.class<#elseif param.mappingTarget>${ext.targetBeanName}<#else>${param.name}</#if><#if param_has_next>,<#else> </#if></#list>)
</@compress>;
<#if hasReturnType()>
if ( ${targetVariableName} != null ) {
    return ${targetVariableName};
}
</#if>

```

filiphr commented:
@pascalgn there is a small bug when someone uses `target` as a parameter name with the mapping methods:
* `MapMappingMethod`
* `IterableMappingMethod`
* `ValueMappingMethod` - didn't actually test
* `EnumMappingMethod` - didn't actually test

You can reproduce it with if you add the following methods to `NumberMapperWithContext`:
```java
public abstract Map<String, Integer> integerMap(Map<String, Long> target);

public abstract List<String> setToList(Set<Integer> target);
```

And the following to `NumberMapperContext`:
```java
 @AfterMapping
public static Map<String, Integer> withMap(Map<String, Long> source, @MappingTarget Map<String, Integer> target) {
    return target;
}

@AfterMapping
public static List<String> withList(Set<Integer> source, @MappingTarget List<String> target) {
    return target;
}
```

Can you make sure that this doesn't happen?

sjaakd commented:
I'm looking at the generated code:

```java
   public NodeDTO nodeToNodeDTO(Node node) {
        org.mapstruct.ap.test.callbacks.returning.NodeDTO target = NodeMapperContext.getInstance( node, NodeDTO.class );
        if ( target != null ) {
            return target;
        }

        if ( node == null ) {
            return null;
        }

        NodeDTO nodeDTO = new NodeDTO();

        NodeMapperContext.setInstance( node, nodeDTO );

        nodeDTO.setParent( nodeToNodeDTO( node.getParent() ) );
        nodeDTO.setName( node.getName() );
        List<NodeDTO> list = nodeListToNodeDTOList( node.getChildren() );
        if ( list != null ) {
            nodeDTO.setChildren( list );
        }
        List<AttributeDTO> list_ = attributeListToAttributeDTOList( node.getAttributes() );
        if ( list_ != null ) {
            nodeDTO.setAttributes( list_ );
        }

        NodeMapperContext.cleanup();

        return nodeDTO;
    }

```

I noticed that although the `set` and the `get` method have an `@BeforeMapping` annotation. However, they end up on different spots in the generated code. Most likely because the first takes the source and the other takes the target. Was this a feature that was already present before this PR or is this introduced as part of this PR?

I also noticed the FQN is used. however, I think I saw @filiphr commenting on that already.

filiphr commented:
@sjaakd it is from before. That is due to the `@MappingTarget` in one of the `@BeforeMapping` methods.

There are 2 types of `@BeforeMapping` methods:
* `beforeMappingReferencesWithoutMappingTarget` - this one is called before the initialization of the `@MappingTarget`
* `beforeMappingReferencesWithMappingTarget` - this one is called after the initialization of the `@MappingTarget` with the instantiated object

agudian commented:
> Was this a feature that was already present before this PR or is this introduced as part of this PR?

@sjaakd: As Filip mentioned, it's a feature that was there all along... I'd hope that we have it documented somewhere as well 😁. You can run `CallbackMethodTest`, which tests out all kinds of different cases and asserts on their order.

agudian commented:
@pascalgn Thanks for the fast turnaround with the changes!
As you can see, we've played around a bit with the generated code and came up with a couple of other spots. Would be great if you could take a look there as well.

sjaakd commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -2093,14 +2095,59 @@ public class VehicleMapperImpl extends VehicleMapper {
+If the before/after-mapping method has a return type other than `void`, it will be checked to match the target type of the mapping methods.
+Methods that have matching parameters, but whose return type does not match, will *not* be used. If the method returns a non-null value,
+this value will be returned from the mapping method.
+
```
Selection of before/after-mapping happens in the following order of preference:
1. the return type of the before/after-mapping method matches with the target type of the mapping method and the parameter types of the before/after-mapping method match with the parameter types of the mapping methods.
2. the return type of before/after-mapping method is void and the parameter types of the before/after-mapping method match with the parameter types of the mapping methods.

agudian replied:
Hmm, Sjaak, I actually think the description given by Pascal is a bit easier. Because there actually isn't any preference in the selection - everything that possibly matches is used (with the order being stable). What's new is that *if* the method returns a non-void type, then that type has to be assignable to the return type of the mapping method. Perhaps the sentences in the doc above can be rephrased to not include so many negations, which I guess is what you're aming for?

sjaakd commented:
Hey guys.. saw a piece of free marker code and started to wonder whether this would also work with update methods. So I extended the mapper with 2 update methods.

```java
@Mapper(uses = NodeMapperContext.class )
public abstract class NodeMapperWithContext {
    public static final NodeMapperWithContext INSTANCE = Mappers.getMapper( NodeMapperWithContext.class );

    public abstract NodeDTO nodeToNodeDTO(Node node);

    protected abstract AttributeDTO attributeToAttributeDTO(Attribute attribute);

    public abstract void nodeToNodeDTO(Node node, @MappingTarget NodeDTO nodeDto);

    protected abstract void attributeToAttributeDTO(Attribute attribute, @MappingTarget AttributeDTO nodeDto);
}
``` 

This yields uncompilable code where the life cycle method assigns to a void:

```java
  @Override
    public void nodeToNodeDTO(Node node, NodeDTO nodeDto) {
        void target = NodeMapperContext.getInstance( node, NodeDTO.class );
        if ( target != null ) {
            return target;
        }

        if ( node == null ) {
            return;
        }
       ...
```

agudian commented on processor/src/main/java/org/mapstruct/ap/internal/model/LifecycleCallbackFactory.java:
```diff
@@ -99,16 +113,18 @@ private LifecycleCallbackFactory() {
+                new LifecycleCallbackMethodReference(
+                    candidate,
+                    parameterAssignmentsForSourceMethod.get( candidate ),
+                    method.getReturnType(),
```
This needs to be `method.getResultType()` (that's also the type used to declare the result object type in the generated method).

agudian commented:
> This yields uncompilable code where the life cycle method assigns to a void:

Exactly, that's what I was going for with my comment on `LifecycleCallbackFactory` above.

And I start to think that we should actually do this in case of `void` update methods:
```
    @Override
    public void nodeToNodeDTO(Node node, NodeDTO nodeDto) {
        NodeDTO target = NodeMapperContext.getInstance( node, NodeDTO.class );
        if ( target != null ) {
            return; // stop with the mapping here
        }

        if ( node == null ) {
            return;
        }
```

Why? It's of course a bit of guess-work what the user would intend to express with the callback method in this case, but at least it would be in sync with the behaviour of the non-void mapping methods (with or without `@MappingTarget` parameter).

WDYT?

sjaakd commented:
> Why? It's of course a bit of guess-work what the user would intend to express with the callback method in this case, but at least it would be in sync with the behaviour of the non-void mapping methods (with or without @MappingTarget parameter).

That would be ok I reckon.. Especially when you can pass a mapping target to the lifecycle method... Can you, does it ?

Remains the question: how do we handle both a @MappingTarget icw a return type?

filiphr commented:
>Why? It's of course a bit of guess-work what the user would intend to express with the callback method in this case, but at least it would be in sync with the behaviour of the non-void mapping methods (with or without @MappingTarget parameter).

I agree that it is OK to stop there.

>Remains the question: how do we handle both a @MappingTarget icw a return type?

I say we just return the value that was returned by the method, same as if there was not `@MappingTarget`. If a user does something like this, then I would expect that they would use the returned value and that they will make sure that the passed value is correctly update in their `@BeforeMapping` method

sjaakd commented:
In the light of the discussion above.. should some test cases be added with update methods? I'm thinking of:
1. void mapping method with `@TargetMapping`
2. mapping method with return type and with `@TargetMapping`
3. passing the `@TargetMapping` to the lifecycle method.

pascalgn commented:
Again, thanks for the feedback, especially for pointing out the bugs with update methods and parameters named "target"! There was an extended discussion, but I think I have addressed all remarks. However, it would be great if everyone could make sure if their issues are fixed in the current commit. Thanks!

sjaakd commented:
@pascalgn thanks! I'm going to take a look asap. Thanks for the quick response.

sjaakd commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -2093,14 +2095,60 @@ public class VehicleMapperImpl extends VehicleMapper {
+
 All before/after-mapping methods that *can* be applied to a mapping method *will* be used. <<selection-based-on-qualifiers>> can be used to further control which methods may be chosen and which not. For that, the qualifier annotation needs to be applied to the before/after-method and referenced in `BeanMapping#qualifiedBy` or `IterableMapping#qualifiedBy`.
 
 The order in which the selected methods are applied is roughly determined by their location of definition (although you should consider it a *code smell* if you need to rely on their order):
```
Although existing text: The *code smell* is not exactly true. `@Before` Lifecycle methods depending on the target type are called after construction of the target type. Your example (avoiding object cycles) is even relying on that. I think.

@agudian: I think this part of the manual should be a bit rephrased (not necessarily part of this PR). Also the 'experimental' nature should be removed (shall we discuss this in a hangout?).

@pascalgn for your PR this is OK.

agudian replied:
The methods are split up into three groups: first the before-mapping methods without mapping-target param, then the before-mapping methods with mapping-target params, and finally the after-mapping methods. In that order, and that's part of the API contract.
But if you need to rely on the order *within one of those groups* (e.g. two after-mapping methods), then that's the code smell I was referring to in that section. The order is kept stable in the generation, but you might get different results with different compilers when the `used` mapper is part of another module. But we might want to rephrase it (in a separate PR) if that wasn't clear.

sjaakd replied:
I understand but it does not immediately stand out as clear. We need to address a non expert public. So, therefore see #983.

sjaakd commented:
@pascalgn Looks good.. My previous comments were handled.

Going over the code I spotted two additional issues. If you can still have a look at those, we're in a good shape to merge AFAIC.

@agudian, @filiphr : WDYT?

pascalgn commented:
@sjaakd Sorry, I might have overlooked something, but could you please specify the two additional issues again?

Regarding the method call order, I think the methods can also be arranged differently, in my example:
1. The first method is called first because it has `@Before` and no `@MappingTarget`
2. The second method is called second because it has `@Before` *and* `@MappingTarget`
3. The third method is called third because it has `@After`

I would still consider it a code smell if one relies on the method definition order, because method reordering should not change program logic, in my opinion.

filiphr commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -2093,14 +2095,60 @@ public class VehicleMapperImpl extends VehicleMapper {
+    private EntityManager entityManager;
+
+    @AfterMapping
+    protected <T> T attachEntity(@MappingTarget Object entity) {
```
I am not sure that this will be used by the mapper.

I added: 

```java
@AfterMapping
public static <T> T attachObject(@MappingTarget Object object) {
    return (T) object;
}
```
And the method is not even considered as a candidate, it is filtered out way sooner in the `MethodRetriavalProcessor`.

I think that `isVoid(type)` check in the `MethodRetriavalProcessor#isValidLifecycleCallbackMethod`. Should be removed.

filiphr replied:
@sjaakd, @agudian This should be a valid callback method right?

filiphr replied:
@pascalgn If the `@MappingTarget` is an `Object` it still does not work. In the `MethodMatcher#matchResultType` and `MethodMatcher#matchSourceType` we have some explicit check to see that the the candidate result/source type is not `java.lang.Object`.

Can you change your the example to:

```java
@AfterMapping
public static <T> T attachObject(@MappingTarget T object) {
    return entityManager.merge(object);
}
```
 This would be similar to the test

sjaakd commented:
> Sorry, I might have overlooked something, but could you please specify the two additional issues again?

Comments above in: `LifecycleCallbackMethodReference.ftl` line 32 and `LifecycleCallbackMethodReference.java`

agudian commented on processor/src/main/resources/org/mapstruct/ap/internal/model/LifecycleCallbackMethodReference.ftl:
```diff
@@ -18,4 +18,16 @@
+</@compress>
+<#if hasReturnType()><#nt>
+if ( ${targetVariableName} != null ) {
+    return<#if methodReturnType.name != "void"> ${targetVariableName}</#if>;
```
Pascal uses the exact same logic as is used in BeanMappimgMethod.ftl, soI think this is okay.

agudian commented on processor/src/main/java/org/mapstruct/ap/internal/model/LifecycleCallbackMethodReference.java:
```diff
@@ -37,19 +38,70 @@
         super( method );
         this.declaringType = method.getDeclaringMapper();
         this.parameterAssignments = parameterAssignments;
+        this.methodReturnType = methodReturnType;
```
I think it's the return type of the mapping method from which this lifecycle method is called, right?

sjaakd replied:
Ok.. If that's the case I've got no further comments..

pascalgn commented:
@filiphr Thanks for spotting this mistake in the documentation! I have updated the example, the method should be considered now.

@sjaakd Concerning the method order: `getInstance` does not take a `NodeDTO` instance, only the type (`NodeDTO.class`) so it can be called before the instance is created. It is therefore always called before `setInstance` will be called, even if the method were declared after `setInstance` in the source code.

@sjaakd It is a bit confusing, but `LifecycleCallbackMethodReference` has itself a return type (`getReturnType`, inherited from `MappingMethod`) and additionally stores the return type *of the method from which that callback method is being called* (`getMethodReturnType`). I am not too happy with the naming, so if you have an idea for better names, just let me know!

Once again, thank you all for the continuously positive discussion and feedback!

filiphr commented:
@pascalgn I am not sure if that was a mistake in your documentation, because this works:

```java
@AfterMapping
public static void attachObject(@MappingTarget Object object) {
    return (T) object;
}
```

I think that the other one should work as well. @sjaakd, @agudian WDYT?

sjaakd commented:
> I am not too happy with the naming, so if you have an idea for better names, just let me know!

Clear. I don't have a better alternative right now.. So lets go ahead.


> Once again, thank you all for the continuously positive discussion and feedback!

No problem. Thanks for your patience 👍

pascalgn commented:
@filiphr There was still a check for return type `void` in `MethodRetrievalProcessor` which I have overlooked so far. I have fixed this, edited the documentation and added another test case. Thanks for being persistent with this issue!

filiphr commented:
@pascalgn Thanks for fixing the issue. There is only one small problem in the example. See my comment there. 

>Thanks for being persistent with this issue!

Thank you for bearing with me 😄. I think that this particular example might be used quite frequently in normal production environment, so I think it is good to make sure that it actually works 😄.

pascalgn commented:
@filiphr I think that `java.lang.Object` should also be supported, but it is explicitly checked to make sure it's not `Object`, so there is probably some valid reasoning behind it. However, I fixed the example to match the test case. Thanks for the input!

filiphr commented:
@pascalgn I am not sure what exactly needs to be done for `java.lang.Object`. However, I think that is not in the scope of this PR and Issue.

@agudian, @sjaakd the PR is OK for me.

agudian commented:
Looks good to me as well, including the handling in void update methods.

@pascalgn I've merged it! 👍 I also added a commit to add you to the list of contributors in the `copyright.txt`.

Thanks a lot for this great feature! Come back any time you like 😄. If you want to, you could also convert your test case to an example project for our mapstruct/mapstruct-examples repository (similar to this mapstruct/mapstruct-examples#12).

AntwaneB commented:
Awesome, thanks a lot for the feature :)

pascalgn commented:
Thanks for all the feedback and thanks to everyone for this positive collaboration!
